src/schedulers/dynamic/scheduler.c

```diff
@@ -274,27 +285,32 @@ void Scheduler_run(Scheduler *untyped_self) {
   bool next_event_is_system_event = false;
   LF_DEBUG(SCHED, "Scheduler running with keep_alive=%d", self->super.keep_alive);
 
-  while (self->super.keep_alive || !self->event_queue->empty(self->event_queue) || untyped_self->start_time == NEVER) {
-    next_tag = self->event_queue->next_tag(self->event_queue);
+  env->enter_critical_section(env);
+
+  while (self->super.keep_alive || !self->event_queue->empty_locked(self->event_queue) ||
+         untyped_self->start_time == NEVER) {
+    next_tag = self->event_queue->next_tag_locked(self->event_queue);
 
     // Check that next tag is greater than start tag. Could be violated if we are scheduling events when the start
     // tag is decided, or receive an input in that period.
     if (lf_tag_compare(next_tag, start_tag) < 0) {
       LF_WARN(SCHED, "Dropping event with tag " PRINTF_TAG " because it is before start tag " PRINTF_TAG, next_tag,
               start_tag);
       ArbitraryEvent e;
-      self->event_queue->pop(self->event_queue, (AbstractEvent *)&e);
+      self->event_queue->pop_locked(self->event_queue, (AbstractEvent *)&e);
       e.event.trigger->payload_pool->free(e.event.trigger->payload_pool, e.event.super.payload);
       continue;
     }
 
     if (env->poll_network_channels) {
+      env->leave_critical_section(env);
```

Comment:
I am really not a big fan of this pattern.

what if `leave_critical_section` was called without being inside a critical section, then `enter_criticial_section` would enter even one, even if we originally were not inside a critical section

Reply:
Calling `leave_critical_section` outside a critical section is a critical error/bug. Many of the platforms will crash the program if that happens